Add reopenable files ext support to the client stats log

gavin-norman-sociomantic opened the pull request.

gavin-norman-sociomantic commented:
Pinged the people who had the application-side problem related to usage of swarm's `ClientStats`.

mihails-strasuns-sociomantic commented on src/swarm/util/log/ClientStats.d:
```diff
@@ -146,6 +152,40 @@ public class ClientStats : StatsLog
+            auto stream = new File(file, File.WriteAppending);
+
+            if ( auto reopenable_files_ext =
+                (cast(Application)app).getExtension!(ReopenableFilesExt) )
```
This is a bit strange - why take `IApplication` only to immediately cast it to`Application`? Why not require `Application` argument at once?

gavin-norman-sociomantic replied:
Yeah true. From the POV of the normal usage, either would be fine, but accepting `Application` directly simplifies this code. Will update.

gavin-norman-sociomantic replied:
(I copied some of this code from another location.)

gautam-kotian-sociomantic commented on src/swarm/util/log/ClientStats.d:
```diff
@@ -146,6 +152,40 @@ public class ClientStats : StatsLog
 
+    /***************************************************************************
+
+        Construct the class to use the default StatsLog appender, registers it
```
Maybe use `Constructs` to be consistent with the tense in the rest of the sentence?

gautam-kotian-sociomantic commented on src/swarm/util/log/ClientStats.d:
```diff
@@ -146,6 +152,40 @@ public class ClientStats : StatsLog
+    /***************************************************************************
+
+        Construct the class to use the default StatsLog appender, registers it
+        with the applications' reopenable files extension (if present), and
```
From swarm's point of view, there'll only be a single application, so I guess the apostrophe should be left-shifted once?

gautam-kotian-sociomantic commented on src/swarm/util/log/ClientStats.d:
```diff
@@ -146,6 +152,40 @@ public class ClientStats : StatsLog
+        starts the update timer.
+
+        Params:
+            app = application class
```
Interesting that you chose to pass the application instance. Why not just the reopenable files extension?

gavin-norman-sociomantic replied:
The idea was that it's slightly more convenient for the user this way.

Commit:
Add ClientStats ctor to register with app's reopenable files extension

Previously, applications would have to do this manually when constructing a
ClientStats instance. As this is non-trivial, most apps are not doing this,
meaning that their client stats log files are not being reopened after
rotation.

Fixes #224.

gavin-norman-sociomantic commented:
Updated.

nemanja-boric-sociomantic commented on integrationtest/client_stats/main.d:
```diff
@@ -0,0 +1,71 @@
+    auto client_stats = new ClientStats(app, new EpollSelectDispatcher,
+        "client_stats.log");
+    test!("==")(app.rof_ext.files.length, 1);
+    test!("==")(app.rof_ext.files[0].toString(), "client_stats.log");
```
Instead of this you could just do `test!("==")(app.rof_ext.reopenFile("client_stats.log"), true)` and then you can drop your `TestReopenableFilesExt`: https://github.com/sociomantic-tsunami/ocean/blob/d103e885bc2c77370d00311a2560f4be4c9d983b/src/ocean/util/app/ext/ReopenableFilesExt.d#L201

gavin-norman-sociomantic replied:
Ah yeah. Good idea. Only works with ocean v3.6.0+ (swarm v4.x.x is on v3.1.0), but I've updated with multi-versioned code (which can be removed in swarm v5.x.x ;)

nemanja-boric-sociomantic replied:
https://github.com/sociomantic-tsunami/swarm/issues/252

Commit:
Add test for registering ClientStats log with app's reopenable files ext
